HW0/P5/P5.txt

```diff
@@ -1 +1,25 @@
-# Your answers here.
+By: Matheus C. Fernandes
+
+Question 1: 255 seconds
+	If one cashier were counting the add the totals of 256 bags it would take 255 seconds because the first count 1+1=2 would count as one secong and every count thereafter would also count as one second 
+
+Question 2: 34 seconds
+	Yes, assuming they imidiatelly divide up the work evenly, they can deivide the 256 bags by 8 workers. Each worker than has to count 32 bags which would take 31 seconds to count. Then 4 workers can add up the other bags, taking 1 additional second. Then 2 workers take the sum of 4 people, taking 1 additional second. Finally, the 1 worker adds the final sum of two people, during the last 1 second.
+
+Question 3: 8 seconds
+	Assuming there is no communication time between workers, it would take 8 seconds. This starts with having 256 workers and taking each second to add each of the half of the workers namely it takes log2(256)=8.
+	
+Question 4: log2(N), please find attached P5.png
+	It would take infinite number of employees log2(N) seconds to count N bags.
+
+Question 5: 15 seconds
+	Every time there is a sum between two numbers there is one additional second, except for the first case. 
+
+Question 6: **this part was done in colaboration with Thibaut Perol**
+	(each cashier comes and takes 1 bag): 230 seconds
+		picture attached 'P6_1bagperworker-8bags.jpg,P6_1bagperworker-16bags.jpg'
+
+	(each cashier comes and takes 2 bags): 132 seconds
+		picture attached 'P6_2bagperworker-16bags.jpg'
+
+	(Better strategy?): No, for having a system where we count two bags at one time, we cannot do better than handing each person two bags at a time. If they were able to count more than two bags at a time, then we would be able to change this strategy.
```

Comment:
This doesn't make sense to me. I will copy my answer:

The most efficient strategy is to distribute bags as fast as
possible. It only takes t_hand to hand *any number of bags* to an employee.
Therefore, we should employee another binary tree method to hand bags
out to employees as fast as possible. Sketching the idea,

      N
    |   |
   N/2 N/2
|    |   |   |
N/4 N/4 N/4 N/4

etc. You start with N bags and hand half to the next employee.
That employee hands half to another employee, and this repeats until
each employee has 1 bag. At this point, employees sum down the binary
tree via communicating with each other. The solution will then scale as
Log2[N] which is quite fast. We thus find

t_minimum = Log2[N] t_hand + Log2[N](t_add + t_communicate)

Simplifying,

t_minimum = Log2[N](t_hand + t_add + t_communicate)

Evidently, the fastest handing strategy results in 3 times the original
infinite solution (without communication lag). Plugging in numbers,
we find

t_minimum = 24 seconds